File: trunk/examples/UI_Analysis/shell_sns.pl
#!/usr/bin/perl

my $job_name   = "pl";

use File::Path;
mkpath(["./out/" . $job_name], 1, 0755);

# ------------------------------------------------------
# Create the ALE::UI::Shell  instance 
# ------------------------------------------------------

use lib ("$ENV{UAL_EXTRA}/ALE/api");
use ALE::UI::Shell;

print "Create the ALE::UI::Shell instance", "\n";

my $shell = new ALE::UI::Shell("print" => "./out/" . $job_name . "/log");

# ------------------------------------------------------
# Define the space of Taylor maps
# ------------------------------------------------------

print "Define the space of Taylor maps", "\n";

$shell->setMapAttributes("order" => 5);

# ------------------------------------------------------
# Define design elements and beam lines
# ------------------------------------------------------

# Read MAD input file

print "Read MAD input file", "\n";

$shell->readMAD("file" => "./data/ff_sext_latnat.mad");

# Split generic elements into thin multipoles

$shell->addSplit("elements" => "^(q[df]h|q[fd][lmc]h|qfbh)\$", 
                 "ir" => 2); 
$shell->addSplit("elements" => "^(bnd)\$", 
                 "ir" => 2); 

# Define aperture parameters: shape, xsize, and ysize.

print "Define aperture parameters", "\n";

# half dipoles

$shell->addAperture("elements" => "^(bnd)\$", 
                    "aperture" => [1., 0.116, 0.079]);

# half quads

$shell->addAperture("elements" => "^(q[fd]h|qdmh)\$", 
                    "aperture" => [1., 0.105, 0.105]);


$shell->addAperture("elements" => "^(qfbh)\$", 
                    "aperture" => [1., 0.13, 0.13]);

$shell->addAperture("elements" => "^(q[fd][lc]h)\$", 
                    "aperture" => [1., 0.15, 0.15]);


$shell->addAperture("elements" => "^(s[13][fd])\$", 
                    "aperture" => [1., 0.105, 0.105]);


$shell->addAperture("elements" => "^(s[24][fd])\$", 
                    "aperture" => [1., 0.13, 0.13]);


# ------------------------------------------------------
# Select and initialize a lattice for operations
# ------------------------------------------------------

print "Select and initialize a lattice", "\n";

# Select an accelerator for operations

$shell->use("lattice" => "ring");

# Write SMF into the FTPOT file

$shell->writeFTPOT("file" => "./out/" . $job_name . "/tpot");

# ------------------------------------------------------
# Write the lattice state to the SXF file
# ------------------------------------------------------

use lib ("$ENV{UAL_EXTRA}/ADXF/api");
use UAL::ADXF::Parser;

my $adxf_parser = new UAL::ADXF::Parser();


$adxf_parser->write("./out/" . $job_name . "/ff_sext_latnat.adxf");

# ------------------------------------------------------
# Define beam parameters
# ------------------------------------------------------

print "Define beam parameters", "\n";

$shell->setBeamAttributes("energy" => 1.93827231, "mass" => 0.93827231);

# ------------------------------------------------------
# Linear analysis
# ------------------------------------------------------

print "Linear analysis: ", "\n";

# Make general analysis
print " analysis\n";

my $dp;
for($dp = -0.02; $dp <= 0.02; $dp += 0.005){
  $shell->analysis("print" => "./out/" . $job_name . "/analysis" . "." . $dp, "dp/p" => $dp); 
}

# Make linear matrix
print " matrix\n";

$shell->map("order" => 1, "print" => "./out/" . $job_name . "/map1"); 

# Calculate survey
print " survey\n";

$shell->survey("elements" => "", "print" => "./out/" . $job_name . "/survey"); 

# Calculate twiss
print " twiss\n";

$shell->twiss("elements" => "", "print" => "./out/" . $job_name . "/twiss"); 

print "End", "\n";

1;

File: trunk/examples/UI_Analysis/shell_sns.cc

#include "UAL/UI/Shell.hh"

using namespace UAL;

int main(){

  UAL::Shell shell;

  // ************************************************************************
  std::cout << "\nDefine the space of Taylor maps." << std::endl;
  // ************************************************************************
  
  shell.setMapAttributes(Args() << Arg("order", 5)); 


  // ************************************************************************
  std::cout << "\nBuild lattice." << std::endl;
  // ************************************************************************
  
  shell.readADXF(Args() << Arg("file",  "./data/ff_sext_latnat.adxf"));

  // ************************************************************************
  std::cout << "\nAdd split ." << std::endl;
  // ************************************************************************

  shell.addSplit(Args() << Arg("lattice", "ring") << Arg("types", "Sbend")
		 << Arg("ir", 2));

  shell.addSplit(Args() << Arg("lattice", "ring") << Arg("types", "Quadrupole")
		 << Arg("ir", 2));

  // ************************************************************************
  std::cout << "Select lattice." << std::endl;
  // ************************************************************************

  shell.use(Args() << Arg("lattice", "ring"));

  // ************************************************************************
  std::cout << "\nWrite ADXF file ." << std::endl;
  // ************************************************************************
   
  shell.writeADXF(Args() << Arg("file",  "./out/cpp/ff_sext_latnat.adxf"));


  // ************************************************************************
  std::cout << "\nDefine beam parameters." << std::endl;
  // ************************************************************************

  shell.setBeamAttributes(Args() << Arg("energy", 1.93827231) 
			  << Arg("mass", 0.93827231));

  // ************************************************************************
  std::cout << "\nLinear analysis." << std::endl;
  // ************************************************************************
  
  // Make linear matrix
  std::cout << " matrix" << std::endl;

  shell.map(Args() << Arg("order", 1) << Arg("print", "./out/cpp/map1"));
  
  // Calculate twiss
  std::cout << " twiss (ring )" << std::endl;
  
  shell.twiss(Args() << Arg("print", "./out/cpp/ring.twiss")); 

  std::cout << " twiss (transfer line)" << std::endl;

  PacTwissData tw;
  tw.beta(0, 2.626);
  tw.alpha(0, 0.5705);
  tw.d(0, 1.406e-05);
  tw.dp(0, 0);

  tw.beta(1, 12.32);
  tw.alpha(1, -2.26);

  
  shell.twiss(Args() << Arg("print", "./out/cpp/tline.twiss"), tw); 


}

